### PR TITLE
Feature/refactor grommet typescript

### PR DESCRIPTION
**Pull request by murbanowicz**
Ignore TS files in ESLint
Add TypeScript with config to enable better support in IDEs
Add TSLint with config to enable better support in IDEs
Create types/common.d.ts to store types used across the whole project
Replace duplicated hard-coded types by common types in all components.

### Summary

I came up with this PR as the start of the work on improving TypeScript support for Grommet.
Adding TSLint/TS and ignoring `*.ts` in ESLint is required to get better support in IDEs when working on Grommet typings and hopefully will be useful in the future in case of rewriting the whole Grommet in TypeScript.

I have found crazy many code repetition in types across many components, like sizes, margins and any function, so I also created common types and replaced those common values. I hope you share the opinion that it makes code much cleaner and easier to read.

Any comments/commits welcome!

**Conflict probably coming from `yarn.lock` as git does not handle merging it efficiently**

### CODE REVIEW

**Comment by ericsoderberghp** [comment]
Are you thinking that we should abandon the `react-desc` way of generating the typescript definitions and switch to hand editing each components file using common utility functions and variables where possible? I think this is what you are proposing but wanted to check.

If so, I am leaning this direction as well. I want to discuss a bit more with @alansouzati first as he did the initial react-desc approach.

Thanks for participating in this way, it helps advance things for everyone.

**Comment by oorestisime** [comment]
another solution which seems to be the "default" way to do in the TS community is having these definitions at http://definitelytyped.org/ https://github.com/DefinitelyTyped/DefinitelyTyped

They seem to be handling an enormous amount of TS typings. This would delegate the work elsewhere and enable you to focus on other things (given that there seems to be typings issues popping often)

OTOH that repo looks huge and maybe that would refrain people from contributing there, and maybe typings will find themselves out of sync.

**Comment by murbanowicz** [comment]
The trend is other way around - people either rewrite in TS or include typings inside the repo as Grommet does so it is not good idea.

**Comment by oorestisime** [comment]
oh ok. not really familiar with all these, just saw that styled components and a few other libraries did that recently.

**Comment by patricknick** [comment]
I don't think moving the types over to DefinitelyTyped would be too bad. As long as there is some community to keep the types more or less in sync with the Grommet library. This approach was also recommended here: https://github.com/styled-components/styled-components/issues/1778#issuecomment-401974135

Surely, at the moment type support is not ideal. Generating the types from PropTypes is in my opinion too generic. Many types are so unclear, quite often we need to cast them as 'any', which is not the idea of typings. Take the select component for example.

The `value` property ends up with this:
```
string | (string & string[]) | (object & string[]) | (JSX.Element & string) | (JSX.Element & number) | (JSX.Element & string[]) | ((string | object)[] & string) | ((string | object)[] & number) | ((string | object)[] & string[]) | undefined
```
Basically, all I can pass in is a string or nothing. The rest is excluded by the union types.

Meanwhile, the onChange function is defined as such:
```
(((...args: any[]) => any) & ((event: React.ChangeEvent<HTMLSelectElement>) => void)) | undefined
```
It's like an `AnyFunction`.

When working with Grommet and Typescript I'd wish for an API, that helps me code faster without having to look at the documentation every time to see what properties I can pass or expect.

Therefore I agree with @ericsoderberghp to abandon react-desc. If on DefinitelyTyped or within Grommet is more of a detail question for me... But I assume rewriting the entire library in Typescript is just not an option.

What does the core team think?

**Comment by oorestisime** [comment]
now that react-desc is out of the equation for TS files i wonder whether it makes sense to rebase this PR in current master and advance with it. it has a lot of fixes that are nice to have i guess.

cc @alansouzati @ericsoderberghp

**Comment by ericsoderberghp** [comment]
I am amenable to that @oorestisime.

**Comment by murbanowicz** [comment]
@oorestisime I am not sure if  owners want to merge it?

**Comment by oorestisime** [comment]
well last eric's comment seemed to mention to be the case.

**Comment by schettino** [comment]
What do you guys feel about a TS rewrite? The project is well structured, and we can have a transition phase and tackle this with an elaborated approach as well. I'm down for it and can dedicate my time on this if you want to move towards it.

**Comment by oorestisime** [comment]
FYI opened an issue to discuss ts rewrite https://github.com/grommet/grommet/issues/2886

**Comment by ShimiSun** [comment]
Closing this PR since we will be tracking the effort on #2886, thank you all for the collaboration!
